jwst/emicorr/emicorr.py

```diff
@@ -0,0 +1,792 @@
+
+    if save_intermediate_results and save_onthefly_reffile is not None:
+        if readpatt == 'FASTR1':
+            freqs_dict = {readpatt: freqs2correct}
+        else:
+            freqs_dict = {readpatt: {detector: freqs2correct} }
+        on_the_fly_subarr_case = {}
+        on_the_fly_subarr_case[subarray] = {
+            'rowclocks': rowclocks,
+            'frameclocks': frameclocks,
+            'freqs': freqs_dict }
+        freq_pa_dict['subarray_cases'] = on_the_fly_subarr_case
+        mk_reffile(freq_pa_dict, save_onthefly_reffile)
+
+    return output_model
+
+
+def sloper(data):
+    """ Fit slopes to all pix of a ramp, using numerical recipies plane-adding
+     returning intercept image.
+
+    Parameters
+    ----------
+    data : numpy array
+        3-D integration data array
+
+    Returns
+    -------
+    outarray : numpy array
+        3-D integration data array
+
+    intercept: numpy array
+        slope intercept values
+    """
+    ngroups, ny, nx = np.shape(data)
+    frametime = 1.0   # use 1.0 for per-frame slopes, otherwise put a real time here
+    grouptimes = np.arange(ngroups) * frametime
+
+    sxy = np.zeros((ny, nx))
+    sx = np.zeros((ny, nx))
+    sy = np.zeros((ny, nx))
+    sxx = np.zeros((ny, nx))
+
+    for groupcount in range(ngroups):
+        # do the incremental slope calculation operations
+        sxy = sxy + grouptimes[groupcount] * data[groupcount, ...]
+        sx = sx + grouptimes[groupcount]
+        sy = sy + data[groupcount, ...]
+        sxx = sxx + grouptimes[groupcount]**2
+
+    # calculate the final per-pixel slope values
+    outarray = (ngroups * sxy - sx * sy)/(ngroups * sxx - sx * sx)
+
+    # calculate the final per-pixel intercept values
+    intercept = (sxx * sy - sx * sxy)/(ngroups * sxx - sx * sx)
+
+    return outarray, intercept
+
+
+def minmed(data, minval=False, avgval=False, maxval=False):
+    """ Returns the median image of a stack of images, or if there are 2 or less
+    non-zero frames, returns the minimum (or this minimum can be forced).
+
+    Parameters
+    ----------
+    data : numpy array
+        3-D integration data array
+
+    minval : bool
+        Returned image will be the minimum
+
+    avgval : bool
+        Returned image will be the mean
+
+    maxval : bool
+        Returned image will be the maximum
+
+    Returns
+    -------
+    medimg : numpy array
+        Median image of a stack of images
+    """
+
+    ngroups, ny, nx = np.shape(data)
+    medimg = np.zeros((ny, nx))
+
+    for i in range(nx):
+        for j in range(ny):
+            vec = data[:, j, i]
+            u = np.where(vec != 0)
+            n = vec[u].size
+            if n > 0:
+                if n <= 2 or minval:
+                    medimg[j, i] = np.min(vec[u])
+                if maxval:
+                    medimg[j, i] = np.max(vec[u])
+                if not minval and not maxval and not avgval:
+                    medimg[j, i] = np.median(vec[u])
```

Comment:
Is there any risk of input values containing NaN's, which could trip up these simple numpy methods?